# Hermitian Matrices, Unitary Matrices and Tensor Product


# Understanding Complex Hermitian and Unitary Matrices

This tutorial delves into two pivotal concepts in quantum computing and linear algebra: complex Hermitian and Unitary matrices. Both play critical roles in quantum mechanics and quantum computing, underpinning the behavior of quantum systems and the operations performed on them.

## Objectives
- Define and explore the properties of complex Hermitian matrices.
- Understand and implement complex Unitary matrices in Python.
- Highlight their significance and applications in quantum computing.

Prepare to enhance your mathematical toolkit with these foundational concepts, illustrated with Python examples for a hands-on learning experience.



## Complex Hermitian Matrices

A Hermitian matrix, or self-adjoint matrix, is a square matrix that is equal to its own conjugate transpose. This property makes it a natural representation of observable quantities in quantum mechanics.

### Properties
- The eigenvalues of a Hermitian matrix are always real.
- The eigenvectors of a Hermitian matrix corresponding to different eigenvalues are orthogonal.
- Hermitian matrices are extensively used in quantum mechanics to describe physical observables.

### Python Example: Defining a Hermitian Matrix


In [2]:

import numpy as np

# Define a complex Hermitian matrix
hermitian_matrix = np.array([[2+0j, 2-1j], [2+1j, 3+0j]])

# Check if the matrix is Hermitian
is_hermitian = np.allclose(hermitian_matrix, hermitian_matrix.conj().T)

is_hermitian, hermitian_matrix


(True,
 array([[2.+0.j, 2.-1.j],
        [2.+1.j, 3.+0.j]]))


## Complex Unitary Matrices

A Unitary matrix is a square matrix whose conjugate transpose is also its inverse. Unitary matrices preserve the inner product, making them essential for describing quantum gates and evolution in quantum computing.

### Properties
- Preserves the norm of vectors, hence preserving quantum states during transformations.
- The determinant of a Unitary matrix has an absolute value of 1.
- Used to represent reversible quantum operations and time evolution in quantum systems.

### Python Example: Defining a Unitary Matrix


In [ ]:

# Define a complex Unitary matrix
unitary_matrix = np.array([[1/np.sqrt(2), 1/np.sqrt(2)], [1/np.sqrt(2)*1j, -1/np.sqrt(2)*1j]])

# Check if the matrix is Unitary
# np.eye(2) returns the identity of size 2x2
is_unitary = np.allclose(np.dot(unitary_matrix, unitary_matrix.conj().T), np.eye(2))

is_unitary, unitary_matrix


# Tensor product


## Tensor Product for Complex Vectors

The tensor product of two complex vectors creates a new vector in a higher-dimensional space, allowing us to combine quantum states in quantum computing.

### Example


In [1]:

import numpy as np

# Define two complex vectors for the tensor product
v1 = np.array([1+2j, 3+4j])
v2 = np.array([5+6j, 7+8j])

# Calculate the tensor product
tensor_product_v = np.kron(v1, v2)

tensor_product_v


array([ -7.+16.j,  -9.+22.j,  -9.+38.j, -11.+52.j])


## Tensor Product for Complex Matrices

The tensor product for matrices combines individual matrices into a larger matrix, crucial for representing multiple qubit systems in quantum computing.

### Example


In [ ]:

# Define two complex matrices for the tensor product
M1 = np.array([[1+2j, 3+4j], [5+6j, 7+8j]])
M2 = np.array([[9+10j, 11+12j], [13+14j, 15+16j]])

# Calculate the tensor product
tensor_product_m = np.kron(M1, M2)

tensor_product_m


# Exercises on Complex Vector and Matrix Operations

This set of exercises is designed to test your understanding of various concepts related to complex vector and matrix operations, which are foundational in quantum computing. Each exercise presents a concrete case for you to apply what you've learned about Hermitian matrices, Unitary matrices, and tensor products.

NOTE: VERIFY YOUR COMPUTATIONS IN PAPER AND IN THE COMPUTER.

## Exercise 1: Complex Hermitian Matrices

Consider the matrix:

$$ H = \begin{bmatrix} 3 & 2+i \\ 2-i & 1 \end{bmatrix} $$

- Verify if $ H $ is a Hermitian matrix.
- If it is, find its eigenvalues.

## Exercise 2: Complex Unitary Matrices

Consider the matrix:

$$ U = \frac{1}{\sqrt{2}}\begin{bmatrix} 1 & i \\ i & 1 \end{bmatrix} $$

- Verify if $ U $ is a Unitary matrix.
- Compute $ UU^\dagger $ to confirm its Unitarity, where $ U^\dagger $ denotes the conjugate transpose of $ U $.

## Exercise 3: Tensor Product for Complex Vectors

Given the complex vectors:

$$ \mathbf{v} = \begin{bmatrix} 1+i \\ 2-i \end{bmatrix}, \quad \mathbf{w} = \begin{bmatrix} 1-2i \\ 3 \end{bmatrix} $$

Calculate the tensor product $ \mathbf{v} \otimes \mathbf{w} $.

## Exercise 4: Tensor Product for Complex Matrices

Given the matrices:

$$ M_1 = \begin{bmatrix} 0 & 1 \\ 1 & 0 \end{bmatrix}, \quad M_2 = \begin{bmatrix} i & 0 \\ 0 & -i \end{bmatrix} $$

Calculate the tensor product $ M_1 \otimes M_2 $.

## Exercise 5: Modelling quantum computations with vectors and matrices

Using matrices and vectors, implement a model of the Mach/Zehnder interferometer.

![Mach-Zehnder interferometer](images/Mach-Zehnder-Interferometer.png)

## Exercise 6: Composing quantum systems 

Using matrices and vectors, implement a model of the following circuit.

![Mach-Zehnder interferometer](images/Deutsch-Algorithm.png)

Use the following MAtrix for $U_f$:

![Mach-Zehnder interferometer](images/ExampleUf.png)



## Exercise 1: Complex Hermitian Matrices

Consider the matrix:

$$ H = \begin{bmatrix} 3 & 2+i \\ 2-i & 1 \end{bmatrix} $$

- Verify if $ H $ is a Hermitian matrix.

## Code part:

In [ ]:
import numpy as np

hermitian_matrix = np.array([[3, 2+1j], [2-1j, 1]])

is_hermitian = np.allclose(hermitian_matrix, hermitian_matrix.conj().T)

is_hermitian

True

$$ True $$

- If it is, find its eigenvalues.

## Code part:

In [ ]:
import numpy as np

h = np.array([[3, 2+1j],[2-1j, 1]])
eigenvalues, eigenvectors = np.linalg.eigh(h)
print("eigenvalues:", eigenvalues)

eigenvalues: [-0.44948974  4.44948974]


$$ eigenvalues: [-0.44948974 , 4.44948974]

$$

## Exercise 2: Complex Unitary Matrices

Consider the matrix:

$$ U = \frac{1}{\sqrt{2}}\begin{bmatrix} 1 & i \\ i & 1 \end{bmatrix} $$

- Verify if $ U $ is a Unitary matrix.

## Code part:

In [19]:
import numpy as np

unitary_matrix = np.array([[1/np.sqrt(2), 1j/np.sqrt(2)], [1j/np.sqrt(2), 1/np.sqrt(2)]])

is_unitary = np.allclose(np.dot(unitary_matrix, unitary_matrix.conj().T), np.eye(2))

is_unitary

True

$$ True $$

- Compute $ UU^\dagger $ to confirm its Unitarity, where $ U^\dagger $ denotes the conjugate transpose of $ U $.

## Code part:

In [20]:
import numpy as np
u = np.array([[1, 1j], [1j, 1]])
udm = np.conjugate(u.T)
print(np.dot(u, udm))

[[2.+0.j 0.+0.j]
 [0.+0.j 2.+0.j]]


$$ U = \begin{bmatrix} 2 & 0 \\ 0 & 2 \end{bmatrix} $$

## Exercise 3: Tensor Product for Complex Vectors

Given the complex vectors:

$$ \mathbf{v} = \begin{bmatrix} 1+i \\ 2-i \end{bmatrix}, \quad \mathbf{w} = \begin{bmatrix} 1-2i \\ 3 \end{bmatrix} $$

Calculate the tensor product $ \mathbf{v} \otimes \mathbf{w} $.

## Code part:

In [21]:
import numpy as np

v1 = np.array([1+1j, 2-1j])
v2 = np.array([1-2j, 3+0j])

Tensor_product = np.kron(v1, v2)
print(Tensor_product)

[3.-1.j 3.+3.j 0.-5.j 6.-3.j]


$$ \mathbf{v} \otimes \mathbf{w} = \begin{bmatrix} 3 -i  \\ 3+3i \\ -5i \\ 6-3i\end{bmatrix}$$

## Exercise 4: Tensor Product for Complex Matrices

Given the matrices:

$$ M_1 = \begin{bmatrix} 0 & 1 \\ 1 & 0 \end{bmatrix}, \quad M_2 = \begin{bmatrix} i & 0 \\ 0 & -i \end{bmatrix} $$

Calculate the tensor product $ M_1 \otimes M_2 $.

## Code part:

In [22]:
import numpy as np
m1 = np.array([[0, 1],[1, 0]])
m2 = np.array([[1j, 0],[0, -1j]])

Tensor_product = np.kron(m1, m2)
print(Tensor_product)

[[0.+0.j 0.+0.j 0.+1.j 0.+0.j]
 [0.+0.j 0.-0.j 0.+0.j 0.-1.j]
 [0.+1.j 0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.-1.j 0.+0.j 0.-0.j]]


$$ M_1 \otimes M_2 = \begin{bmatrix} 0 & 0 & i & 0  \\ 0 & 0 & 0 & -i  \\ i & 0 & 0 & 0  \\ 0 & -i & 0 & 0 \end{bmatrix}$$

## Exercise 5: Modelling quantum computations with vectors and matrices

Using matrices and vectors, implement a model of the Mach/Zehnder interferometer.

![Mach-Zehnder interferometer](images/Mach-Zehnder-Interferometer.png)

## Code part:

In [23]:
import numpy as np
v0 = np.array([[1],[0]])
H = (1/(np.sqrt(2)))*(np.array([[1, 1],[1, -1]]))
X = np.array([[0, 1],[1, 0]])
v1 = np.dot(H, v0)
v2 = np.dot(X, v1)
v3 = np.dot(H, v2)
print(v1)
print()
print(v2)
print()
print(v3)

[[0.70710678]
 [0.70710678]]

[[0.70710678]
 [0.70710678]]

[[1.00000000e+00]
 [2.23711432e-17]]


$$ \begin{bmatrix} 0.70710678 \\ 0.70710678 \end{bmatrix}, \begin{bmatrix} 0.70710678 \\ 0.70710678 \end{bmatrix}, \begin{bmatrix} 1.00000000e+00 \\ 2.23711432e-17 \end{bmatrix} $$

## Exercise 6: Composing quantum systems 

Using matrices and vectors, implement a model of the following circuit.

![Mach-Zehnder interferometer](images/Deutsch-Algorithm.png)

Use the following MAtrix for $U_f$:

![Mach-Zehnder interferometer](images/ExampleUf.png)

## Code part:

In [25]:
import numpy as np
v1_1 = np.array([[1],[0]])
v1_2 = np.array([[0],[1]])
H = (1/(np.sqrt(2)))*np.array([[1, 1],[1, -1]])
I = np.array([[1, 0],[0, 1]])
HH = np.kron(H, H)

HI = np.kron(H, I)

U = np.array([[0, 1, 0, 0],[1, 0, 0, 0],[0, 0, 1, 0],[0, 0, 0, 1]])
v1_3 = np.kron(v1_1, v1_2)
v1_4 = np.dot(HH, v1_3)
v1_5 = np.dot(U, v1_4)
v1_6 = np.dot(HI, v1_5)

print(v1_3)
print()
print(v1_4)
print()
print(v1_5)
print()
print(v1_6)

[[0]
 [1]
 [0]
 [0]]

[[ 0.5]
 [-0.5]
 [ 0.5]
 [-0.5]]

[[-0.5]
 [ 0.5]
 [ 0.5]
 [-0.5]]

[[ 0.        ]
 [ 0.        ]
 [-0.70710678]
 [ 0.70710678]]


$$ \begin{bmatrix} 0 \\ 1 \\0 \\ 0\end{bmatrix}, \begin{bmatrix} 0.5 \\ -0.5 \\0.5 \\ -0.5 \end{bmatrix}, \begin{bmatrix} -0.5 \\ 0.5 \\0.5 \\ -0.5 \end{bmatrix}, \begin{bmatrix} 0 \\ 0 \\-0.70710678 \\ 0.70710678 \end{bmatrix} $$